compiler/rustc_hir_analysis/src/check/mod.rs

```diff
@@ -319,36 +319,70 @@ fn bounds_from_generic_predicates<'tcx>(
             _ => {}
         }
     }
+
+    let mut where_clauses = vec![];
     let generics = if types.is_empty() {
         "".to_string()
     } else {
-        format!(
-            "<{}>",
-            types
-                .keys()
-                .filter_map(|t| match t.kind() {
-                    ty::Param(_) => Some(t.to_string()),
+        let types_str = types
```

Comment:
This is a bit too convoluted imo...

Can this be restructured into a loop that pushes to multiple vecs?